crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -122,4 +122,116 @@ module sui::random {
     ) {
         update_randomness_state(self, new_round, new_bytes, ctx);
     }
+
+
+    /// Unique randomness generator, derived from the global randomness.
+    struct RandomGenerator has drop {
+        seed: vector<u8>,
+        counter: u64,
+        buffer: vector<u8>,
+    }
+
+    const RAND_OUTPUT_LEN: u16 = 32;
+    const MAX_RANDOM_BYTES: u16 = RAND_OUTPUT_LEN * 100;
+
+    /// Create a generator.
+    public fun new_generator(r: &Random, ctx: &mut TxContext): RandomGenerator {
+        let inner = load_inner(r);
+        // TODO[crypto]: should blake be used here instead?
+        let seed = hmac_sha3_256(&inner.random_bytes,&to_bytes(fresh_object_address(ctx)));
+        RandomGenerator { seed, counter: 0, buffer: vector::empty() }
+    }
+
+    // TODO[move]: Should any of the following functions be implemented as native functions to save gas?
```

Comment:
Best thing to do is to measure and see.  I kind of doubt it will matter, but if it did you could implement the generator for random integers from the random bytes as a generic native function:

```move
native fun gen<T: copy + drop + store>(g: &mut RandomGenerator): T;
```

And thus avoid having to create a U256 every time (even for smaller types) and run the loop in Move.  It will be cheaper, but it's hard to imagine people calling this so much that they'd care.

Reply:
yeah, I like that, though the native function is a bit of an annoyance.
I would not worry about gas here, if anything I would rather check how expensive they are in terms of time

Reply:
Removed that comment since I improved the design and removed some inner loops. Worst case is that we loop when we copy <=32 bytes, or if we generate a very long bytes vector.